### PR TITLE
Add Onnx Supported Layers

### PR DESCRIPTION
**Pull request by LeicongLi**
## What changes were proposed in this pull request?
a. Add ONNX supported operators into BigDL
  1. AveragePool
  2. BatchNormalization
  3. Concat
  4. Conv
  5. Gather
  6. Gemm
  7. MaxPool
  8. Relu
  9. Reshape
  10. Shape
  11. Softmax
  12. OnnxSum
  13. Unsqueeze

b. Add corresponding test, in progress

c. Add its Python API

Most operators above build a 1-to-1  mapping to existing  BigDL layer,
Gemm is compound, consist of existing BigDL layers.
Reshape is new. 

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If it is possible, please attach a screenshot; otherwise, remove this)

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/BatchNormalization.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Carries out batch normalization as described in the paper https://arxiv.org/abs/1502.03167.
+ *
+ * epsilon: float (default is 1e-05). The epsilon value to use to avoid division by zero.
+ * momentum: float (default is 0.9). Factor used in computing the running mean and variance.
+ *             e.g., running_mean = running_mean * momentum + mean * (1 - momentum).
+ */
+object BatchNormalization {
+  def apply[T: ClassTag](
+    numFeatures: Int, // number of input features, BigDL requires.
+    epsilon: Float = 1e-05f,
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/BatchNormalization.scala]: default values  are from onnx?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Concat.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import com.intel.analytics.bigdl.nn
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+
+/**
+ * Concatenate a list of tensors into a single tensor
+ */
+object Concat {
+  def apply[T: ClassTag](
+        nInputDims: Int, // specify the number of dimensions of input, BigDL requires.
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Concat.scala]: so onnx does not have this parameter?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Conv.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * strides: Stride along each spatial axis.
+ */
+object Conv {
+  def apply[T: ClassTag](
+    nInputPlane: Int, // BigDL requires
+    nOutputPlane: Int, // BigDL requires
+    kernelShape: List[Int],
+    weight: Tensor[T], // BigDL requires
+    bias: Tensor[T], // BigDL requires
+    autoPad: String = "NOTSET", // missing in BigDL
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Conv.scala]: what's the meaning?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Gather.scala [hunk 4]
```diff
@@ -17,12 +17,14 @@
...
 import scala.reflect.ClassTag
+import com.intel.analytics.bigdl.nn
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
 
 
 object Gather {
-  def apply[T: ClassTag, D: ClassTag]()(implicit ev: TensorNumeric[T], ev2: TensorNumeric[D]):
-  ops.Gather[T, D] = new ops.Gather()
+  def apply[T: ClassTag, D: ClassTag](
+    axis: Int = 0
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Gather.scala]: add one parameter,but where it is used?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Gemm.scala [hunk 5]
```diff
@@ -43,13 +45,14 @@ class Gemm[T: ClassTag](
   val alpha: Float = 1, val beta: Float = 1,
   val transA: Boolean = false, val transB: Boolean = false
 )(implicit ev: TensorNumeric[T])
-extends Operation[Table, Tensor[T], T] {
+extends AbstractModule[Table, Tensor[T], T] {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Gemm.scala]: you should change it to object

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    ceilMode: Int = 0,
+    dilations: List[Int] = null,
+    pads: List[Int] = null,
+    storageOrder: Int = 0,
+    strides: List[Int] = null
+  )(implicit ev: TensorNumeric[T]): nn.SpatialMaxPooling[T] = {
+
+    val (kW: Int, kH: Int) = kernelShape match {
+      case List(width, height) => (width, height)
+      case _ => throw new IllegalArgumentException("Bad kernel shape value")
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala]: Print the values may help us debug error.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/OnnxSum.scala [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package com.intel.analytics.bigdl.nn.onnx
+
+import scala.reflect.ClassTag
+import com.intel.analytics.bigdl.nn
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+
+object OnnxSum {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/OnnxSum.scala]: Why do you name it as  `OnnxSum,` not `Sum` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/BatchNormalization.scala (hunk 1 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/BatchNormalization.scala]: yes

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Concat.scala (hunk 2 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Concat.scala]: No, it does not.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Conv.scala (hunk 3 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Conv.scala]: The way how it should be padded automatically, it could be "Same, Valid"

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Gather.scala (hunk 4 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Gather.scala]: Currently, BigDL ops.Gather does not support this parameter. In order to support it, we need to enrich the layer or implement another one, which I consider to do in next step.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala (hunk 6 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala]: Sure, will add more debugging info to the exceaption.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/OnnxSum.scala (hunk 7 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/OnnxSum.scala]: Name conflict.
There is a layer called Sum under nn, but it does not do what this Sum does.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Constant.scala [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.intel.analytics.bigdl.nn.onnx
+
+import scala.reflect.ClassTag
+// import com.intel.analytics.bigdl.nn.abstractnn.TensorModule
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Constant.scala]: Code clean.

Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Constant.scala]: @i8run Thanks for your time.
According to your suggestions
1. changed parameter's name
2. removed some unwanted layers, added forward tests for the remainning layers, and those layers do not have backward, e.g. Shape layer, it only depends on the input, gradient does not apply to it, right?
3. cleaned up code.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala [hunk 9]
```diff
@@ -30,17 +31,23 @@ import scala.reflect.ClassTag
...
     (1 to dimSize).foreach(i => {
       output.setValue(i, ev.fromType(input.size(i)))
     })
     output
   }
 
+  // Shape if a constant layer,
+  // which means Output does not depend on Weight which gets updated by Gradient
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    gradInput
```
[the remark this answers is not in the document]
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala]: I do not see why we need gradient in shape layer, since the output depends entirely on input.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+
+object MaxPool {
+  def apply[T: ClassTag](
+    kernelShape: List[Int],
+    autoPad: String = "NOTSET",
+    ceilMode: Int = 0,
+    dilations: List[Int] = null,
+    pads: List[Int] = null,
+    storageOrder: Int = 0,
```
[the remark this answers is not in the document]
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala]: storage_order : int (default is 0)
The storage order of the tensor. 0 is row major, and 1 is column major.

Its a parameter for ONNX, we dont have it in BigDL. For now just add it here as a place holder.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+import scala.reflect.ClassTag
+import com.intel.analytics.bigdl.nn
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+
+object MaxPool {
+  def apply[T: ClassTag](
+    kernelShape: List[Int],
+    autoPad: String = "NOTSET",
+    ceilMode: Int = 0,
```
[the remark this answers is not in the document]
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/MaxPool.scala]: ceilMode's default value is 0 in ONNX, and convert to Boolean afterwards.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala [hunk 12]
```diff
@@ -30,17 +31,23 @@ import scala.reflect.ClassTag
  * @tparam T The numeric type in this module parameters
  */
 class Shape[T: ClassTag](implicit ev: TensorNumeric[T])
-  extends Operation[Tensor[T], Tensor[T], T] {
+  extends TensorModule[T] {
 
   override def updateOutput(input: Tensor[T]): Tensor[T] = {
     val dimSize = input.nDimension()
+    Tensor[T]()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala]: what't this?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLOnnx.scala [hunk 13]
```diff
@@ -16,24 +16,297 @@
...
+      case List(elem) => elem + 1 // Todo
+      case _ => throw new IllegalArgumentException("Bad axes value: " + axes)
+    }
+    OnnxOpsMapper.unsqueeze.apply(pos, numInputDims)
+  }
 }
 
 
-object PythonBigDLOnnx {
+object OnnxOpsMapper {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLOnnx.scala]: Why create this object?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/Cast.scala [hunk 14]
```diff
@@ -45,6 +45,7 @@ class Cast extends TensorflowOpsLoader {
       case DataType.DT_STRING => CastOps[T, String]()
       case DataType.DT_FLOAT => CastOps[T, Float]()
       case DataType.DT_DOUBLE => CastOps[T, Double]()
+      case _ => throw new UnsupportedOperationException()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/Cast.scala]: add some error msg

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/ParseExample.scala [hunk 15]
```diff
@@ -42,6 +42,7 @@ class ParseExample extends TensorflowOpsLoader {
         case DataType.DT_FLOAT => FloatType
         case DataType.DT_DOUBLE => DoubleType
         case DataType.DT_STRING => StringType
+        case _ => throw new IllegalArgumentException
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/ParseExample.scala]: add some error msg

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala (hunk 12 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala]: Thanks. Mistakenly added it.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLOnnx.scala (hunk 13 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLOnnx.scala]: The mapper object you suggested before.

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDLOnnx.scala]: no, I think all your mapping is in `PythonBigDLOnnx`.  `OnnxOpsMapper ` is meaningless.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala]: Because we need chain based backwarding, you need to pass `gradientInput` properly into proceeding layer. it's true the output depends entirely on input, so it would be zero, but expected to have the correct size.

[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/onnx/Shape.scala]: make its value zero explicitly,

**Comment by wzhongyuan** [comment]
Some general comments. @LeicongLi  @i8run  @zhangxiaoli73 , we can discuss if you have different view
1) ONNX support is to load ONNX model into BigDL model, so here the focus is how to turn ONNX model into BigDL model , not to expose ONNX related things to users. from user's perspective, he/she just eventually wants to see a BigDL model. Also users are not expected to define model in ONNX way.
2) We may not want to see a list of ONNX layer definition in Python, what we need to implement is the converter/loader. Typically there are 2 steps :  build an ONNX graph into BigDL graph, convert ONNX ops node into BigDL node.   Torch, Caffe and TF support in BigDL are all done this way, especially please refer to TF support implementation, which is very similar to what we need to do for now.

**Comment by i8run** [comment]
LGTM.

**Comment by LeicongLi** [comment]
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/900/
